src/libopenrave-core/jsonparser/jsonreader.cpp

```diff
@@ -901,12 +1001,266 @@ class JSONReader
+
+                            if (!ok) {
+                                throw OPENRAVE_EXCEPTION_FORMAT("failed to parse json document \"%s\"", currentUri, ORE_InvalidArguments);
+                            }
+                            _CurlCleanup(i);
+                            return;
+                        }
+                    }
+                }
+                curlmsg = curl_multi_info_read(_curlMultiHandle, &msgInQueue);
+            }
+        } while(stillRunning != 0);
+
+    }
+
+    void _DownloadConnectedBodies()
+    {
+        if (!_curlDataVector.size()) {
+            return;
+        }
+
+        std::vector<int> vInputToBodyInfoMapping;
+        // TODO make function to replace it with remote URL
+
+
+        int stillRunning = 0;
+        int msgInQueue = 0;
+        do {
+            CURLMcode mc = curl_multi_perform(_curlMultiHandle, &stillRunning);
+            CURLMsg  *curlmsg = curl_multi_info_read(_curlMultiHandle, &msgInQueue);
+            if(!mc && stillRunning) {
+                /* wait for activity, timeout or "nothing" */
+                mc = curl_multi_poll(_curlMultiHandle, NULL, 0, 1000, NULL);
+            }
+            if(mc) {
+                break;
+            }
+
+            while(curlmsg != NULL) {
+                if (curlmsg->msg == CURLMSG_DONE) {
+                    for ( unsigned long i = 0; i < _curlDataVector.size(); i++) {
+                        if (_curlDataVector.at(i).get()->curl == curlmsg->easy_handle) {
+                            boost::shared_ptr<rapidjson::Document> sharedDocument = boost::shared_ptr<rapidjson::Document>(new rapidjson::Document);
+                            rapidjson::ParseResult ok = sharedDocument.get()->Parse<rapidjson::kParseFullPrecisionFlag>(_curlDataVector.at(i).get()->buffer->GetString());
+
+                            if (!ok) {
+                                throw OPENRAVE_EXCEPTION_FORMAT("failed to parse json document \"%s\"", _curlDataVector.at(i).get()->uri, ORE_InvalidArguments);
+                            }
+
+                            _PutDocumentIntoRapidJSONMap(_curlDataVector.at(i).get()->uri, sharedDocument);
+                            _ParseDocumentForNewURLs( *sharedDocument);
+
+                            _CurlCleanup(i);
+
+                            i--;
+                            stillRunning++; // Increase still running by 1 to start new curl downloads even if the other ones have finished
+                        }
+                    }
+                }
+                curlmsg = curl_multi_info_read(_curlMultiHandle, &msgInQueue);
+            }
+        } while(stillRunning != 0);
+
+        return;
+    }
+
+
+
+    /// \brief data type that is passed to curl handler containing information on parsing incoming files
+    struct curlData
+    {
+        rapidjson::StringBuffer *buffer = new rapidjson::StringBuffer(); // Internal buffer assigned once curl is set up to download
```

Comment:
This is a leak waiting to happen.

Reply:
I have not implemented the rotating string buffers, i however have the .hpp file from a previous implementation